Migrate multi dimensional configs to new storage

rakyi opened the pull request:
* Add new HTTP API for upserting mdim configs
* Configs are normalized on creation
  * Indicator paths are resolved to variable IDs - this was previously done when baking
  * We create a full chart config for every view by merging config defined in the view with variable config already stored in the DB
  * We add the chart config UUID for each config to the `view` in the multidim config
  * For each view, we create a row in the newly added `multi_dim_x_chart_configs` table, to track the relationships between multidim views and chart configs  
* To migrate the existing configs, we have to run the `devTools/migrateMultiDimConfigs/migrateMultiDimConfigs.ts` script with:

  ```
  yarn tsx --tsconfig tsconfig.tsx.json devTools/migrateMultiDimConfigs/migrateMultiDimConfigs.ts
  ```

  The existing config must be valid for the migration to pass.
* Use the new chart configs in the multidim data page

In follow-up PRs, we'll:
* Update mdim config and related view chart configs when indicator configs are updated in admin/ETL
* Trigger baking of existing mdim page when its config or config it depends on gets updated
* Enable embedding of mdim views, making use of the new full chart configs and mdim configs stored in R2

rakyi commented:
It looks like `GRAPHER_CONFIG_R2_BUCKET` and `GRAPHER_CONFIG_R2_BUCKET_PATH` are not configured in the staging server. Should we add them to the template @danyx23?

danyx23 reviewed (commented):
@rakyi if we do that then the staging servers will all start to upload modified chart configs to r2 in the staging bucket and we don't have a cleanup there atm. It's maybe best to talk about the pros and cons of doing this different ways in a brief call?

rakyi commented on packages/@ourworldindata/types/src/dbTypes/MultiDimXChartConfigs.ts:
```diff
@@ -0,0 +1,9 @@
+export const MultiDimXChartConfigsTableName = "multi_dim_x_chart_configs"
+
+export type DbInsertMultiDimXChartConfig = {
+    multiDimId: number
```
Since it's an auto-increment field, I think it makes sense to explicitly not have the `id`, because including it is invalid, and you must leave it out. Which is different from optional fields, like `updatedAt`.

rakyi commented on db/model/Variable.ts:
```diff
@@ -141,6 +122,29 @@ export async function getMergedGrapherConfigForVariable(
+    knex: db.KnexReadonlyTransaction,
+    variableIds: number[]
+): Promise<Map<number, GrapherInterface>> {
+    // FIXME: Optimize to a single query that only fetches the data we need.
```
It would be, but there isn't enough time, at least not in this PR.

rakyi commented on adminSiteServer/migrateMultiDimConfigs.ts:
```diff
@@ -0,0 +1,29 @@
+ * This isn't an actual DB data migration because of circular dependencies
+ * between `adminSiteServer/multiDim` and `db`.
+ */
+async function main() {
```
Will move to `devTools`.

It could be a DB data migration (as written), not a DB schema migration. Could you say more about what would you change about the wording?

rakyi commented on adminSiteServer/multiDim.ts:
```diff
@@ -0,0 +1,205 @@
+                ...variableConfigs.get(variableId),
+                ...view.config,
+                dimensions: MultiDimDataPageConfig.viewToDimensionsConfig(view),
+                selectedEntityNames: config.defaultSelection ?? [],
```
This is what we've been doing for mdims on the frontend, and I just copied it here. @marcelgerber do you know why we do this?

rakyi replied:
Seems to work fine and keep the selection in practice, even though it seems like it should not.

marcelgerber replied:
Yes 😁 that's what I also found, and I also don't know _exactly_ why it works, but in reality it does!

rakyi commented:
I added another commit which handles updates to mdim config more granularly.

I changed the DB schema, so if you already ran the migrations locally you should revert them and then run them again. Afterward you'll need to also run the migration script again, which now also changed location, with:

```
yarn tsx --tsconfig tsconfig.tsx.json devTools/migrateMultiDimConfigs/migrateMultiDimConfigs.ts
```

rakyi commented on adminSiteServer/multiDim.ts:
```diff
@@ -0,0 +1,205 @@
+    const enrichedViews = await Promise.all(
+        config.views.map(async (view) => {
+            const variableId = view.indicators.y[0]
+            const fullGrapherConfig = {
```
Good catch, changed. We now get this warning: "About to merge Grapher configs with missing schema information". That's because the mdim config doesn't have a `$schema` key. Should we add it in ETL?

rakyi replied:
Also, the last config we merge in and define inline doesn't have a `$schema`. Should we specify it there too? Hard-coding it to the latest?

marcelgerber commented on db/migration/1729600015045-AddMultiDimIdConfigMd5SlugIndex.ts:
```diff
@@ -0,0 +1,30 @@
+        await queryRunner.query(`-- sql
+            ALTER TABLE multi_dim_data_pages
+            DROP PRIMARY KEY,
+            ADD COLUMN id SERIAL PRIMARY KEY FIRST,
```
I looked up `SERIAL` because I wasn't familiar with it, and it turns out it is a bit funky: https://stackoverflow.com/a/65926920

It's possibly better to use this (same in the other migration):

```suggestion
            ADD COLUMN id INT unsigned NOT NULL AUTO_INCREMENT PRIMARY KEY FIRST,
```

rakyi replied:
Good catch, I should have checked more thoroughly when Copilot generated it. Changed.

marcelgerber reviewed (approved):
Nice! Well done, all good from my side.

rakyi commented on adminSiteServer/multiDim.ts:
```diff
@@ -0,0 +1,348 @@
+    )
+    for (const { viewId, chartConfigId } of rows) {
+        mapping.set(viewId, chartConfigId)
+    }
```
Changed.

danyx23 commented on db/model/MultiDimDataPage.ts:
```diff
@@ -1,10 +1,25 @@
     DbEnrichedMultiDimDataPage,
 } from "@ourworldindata/types"
 
+/** Returns ID of the upserted row or zero, if there are no updates. */
```
I don't understand in which case this would return zero. Can you elaborate?

rakyi replied:
I reworded the doc. Is it clear now?

danyx23 replied:
Yes it is - but this behaviour isn't very obvious to me from the docs. Have you verified that this really returns 0 if there is no difference in the upserted data? I'm very happy to run with this if you did but I'd like to avoid a codepath that we add that, for whatever reasons, in practice doesn't happen like this.

rakyi replied:
I did test it. That's how I found about it. I just double-checked and yep, that's how it works.

danyx23 replied:
Perfect, thanks! Good to go then!

danyx23 reviewed (commented):
I have one question to clarify - once I understand this part this is good to go!

danyx23 reviewed (approved):
👍
